README


This archive contains the distribution CPAN-Repository,
version 0.008:

  API to access a directory which can be served as CPAN repository

This software is copyright (c) 2012 by DuckDuckGo, Inc. L<http://duckduckgo.com/>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
==================================================
Changes from 1739-06-13 00:00:00 +0000 to present.
==================================================

------------------------------------------
version 0.008 at 2013-03-27 03:45:29 +0000
------------------------------------------

  Change: 751bb68a9cc2167f37f9e7ba3e122732cbcfec34
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-02-19 05:09:09 +0000

    Added possibility for custom author path, tuned tests a bit 

  Change: 826b2b1016705a4dd35664ed2d38e1378d4b3377
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2012-07-09 17:22:08 +0000

    A bit more handles on CPAN::Repository, a bit more informational
    functions 

  Change: d0fd8a4a28dfc85f2c5d00d7e07484a49bc49ff9
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2012-07-09 17:09:21 +0000

    Hups, missed generation a bit ;) 

  Change: 817fc4ca199e2ed86904a541aea4d3f25f1fdc42
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2012-07-09 17:06:57 +0000

    Added 06perms.txt 

  Change: 638fbb234a31dbad87c44dd518b05edbffedf9de
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2012-07-09 16:29:41 +0000

    Added sorted output to 02packages file 

  Change: 40882f1e46e8a07633387632621ef6958b9989d6
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-26 01:33:00 +0000

    v0.007 

  Change: 07f6d26a55d513de7cc6a3fa0201086e3885905a
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-26 01:31:50 +0000

    Damn it, remove debugging before release! 

  Change: fc73dd923155974b0ecbaaa1a392805de467dbf1
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-26 01:23:36 +0000

    v0.006

    Fixed handling of undef versions in generation of packages 

  Change: 4a55e90e39dee509b9dfc520f187b77a575fa46e
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-26 01:07:32 +0000

    Fixed handling of undef versions in generation of packages 

  Change: 6805f253673872c2c33d7e7a61049401b5cca191
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-19 18:27:21 +0000

    v0.005

    Return pathname in repo on add_author_distribution 

  Change: 686c92891b684192140ed705ceb8748600bf87b6
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-19 18:25:58 +0000

    Return pathname in repo on add_author_distribution 

  Change: 282f22ce49f2842a23a144cef129d58d16aa463c
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-18 01:01:32 +0000

    v0.004

    Fixed documentation, added generation of 02STAMP 

  Change: 2fcd4dea3741769ccd95fe5c423340bc1d0566f1
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-18 01:00:34 +0000

    Added timestamp feature, added more pod (thats always good) 

  Change: 222fb80936d073994ae6c0cba3a9a1b0ed509183
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-17 04:17:02 +0000

    Fixed typo in SYNOPSIS and a stupid 1; at the end of pod 

  Change: 20f044868e11d112bc06a9e6112e239f3b438831
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-17 04:12:15 +0000

    v0.003

    Arghl... another requirement fix 

  Change: 31a77af4edb34f830ec8be8c68b637f186a56bfa
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-17 04:11:09 +0000

    And again forgot Changes... arghl^2 

  Change: f6dd17fb84ba3d3b2531e46dbbf73f56513da665
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-17 04:10:49 +0000

    Arghl.. another requirement fix... 

  Change: d9fcfd51a0ff36f5aa5a98c663f6a4718b0af712
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-17 04:10:38 +0000

    v0.002

    Fixed requirements 

  Change: 175d5a7e480fb0b1b18d8ba6679e725a3654407d
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-17 04:08:55 +0000

    Updated changes file 

  Change: 2a3ca1d5ffd28f0f67b87ffdb53957c604c8670d
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-17 03:52:46 +0000

    Fixing requirements 

  Change: 133b7ed23dba68b6ff2244c31a344131645a6224
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-17 03:51:54 +0000

    v0.001

    First working version 

  Change: 409bfa2da2cf92ae4c40091908f80c848446642f
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-17 03:48:20 +0000

    Avoid a Moo problem 

  Change: 7a8ae7b59f831360bd04086e858ca3532581a29b
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-17 03:15:25 +0000

    added modules function to CPAN::Repository 

  Change: 25dac06b010a618281df78cfb9fc736229590c97
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-17 02:49:12 +0000

    Fixed reference to old distribution name, added other modules to load
    test 

  Change: b8c0a8be2e3be695c7b1e087ee91a89bc27c84fc
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-17 02:45:56 +0000

    Hups, Data::Dumper probably unrequired ;) 

  Change: 7a3562d96831d9eead9c7ddd188b25de62a34af5
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-17 02:45:06 +0000

    First working version 

  Change: a154fad7f10a8c08ae06ae6daa0d101f0955058b
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-16 01:38:42 +0000

    initial commit 

================
End of releases.
================


Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "API to access a directory which can be served as CPAN repository",
  "AUTHOR" => "Torsten Raudssus <torsten\@raudss.us> L<http://raudss.us/>",
  "BUILD_REQUIRES" => {
    "File::Temp" => "0.22",
    "Test::LoadAllModules" => "0.021",
    "Test::More" => "0.96"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "CPAN-Repository",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "CPAN::Repository",
  "PREREQ_PM" => {
    "DateTime" => "0.72",
    "DateTime::Format::Epoch" => "0.13",
    "DateTime::Format::RFC3339" => 0,
    "Dist::Data" => "0.002",
    "File::Path" => "2.08",
    "File::Spec::Functions" => "3.33",
    "IO::File" => "1.14",
    "IO::Zlib" => "1.10",
    "Moo" => "0.009013"
  },
  "VERSION" => "0.008",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/CPAN/Repository.pm
package CPAN::Repository;
BEGIN {
  $CPAN::Repository::AUTHORITY = 'cpan:GETTY';
}
{
  $CPAN::Repository::VERSION = '0.008';
}
# ABSTRACT: API to access a directory which can be served as CPAN repository

use Moo;
use File::Path qw( make_path );
use File::Spec::Functions ':ALL';
use CPAN::Repository::Mailrc;
use CPAN::Repository::Packages;
use CPAN::Repository::Perms;
use File::Copy;

our $VERSION ||= '0.0development';

has dir => (
	is => 'ro',
	required => 1,
);

has real_dir => (
	is => 'ro',
	lazy => 1,
	builder => '_build_real_dir',
);

sub _build_real_dir { catdir(splitdir(shift->dir)) }

sub splitted_dir { splitdir(shift->real_dir) }

has url => (
	is => 'ro',
	lazy => 1,
	builder => '_build_url',
);

sub _build_url { 'http://cpan.perl.org/' }

has written_by => (
	is => 'ro',
	lazy => 1,
	builder => '_build_written_by',
);

sub _build_written_by { (ref shift).' '.$VERSION }

has mailrc => (
	is => 'ro',
	lazy => 1,
	builder => '_build_mailrc',
	handles => [qw(
		get_alias
	)],
);

sub _build_mailrc {
	my ( $self ) = @_;
	return CPAN::Repository::Mailrc->new({
		repository_root => $self->real_dir,
	});
}

has perms => (
	is => 'ro',
	lazy => 1,
	builder => '_build_perms',
	handles => [qw(
		get_perms
		get_perms_by_userid
	)],
);

sub _build_perms {
	my ( $self ) = @_;
	return CPAN::Repository::Perms->new({
		repository_root => $self->real_dir,
		written_by => $self->written_by,
	});
}


has packages => (
	is => 'ro',
	lazy => 1,
	builder => '_build_packages',
	handles => [qw(
		get_module
		get_module_version
	)],
);

sub _build_packages {
	my ( $self ) = @_;
	return CPAN::Repository::Packages->new({
		repository_root => $self->real_dir,
		url => $self->url,
		written_by => $self->written_by,
		authorbase_path_parts => [$self->authorbase_path_parts],
	});
}

sub is_initialized {
	my ( $self ) = @_;
	$self->mailrc->exist && $self->packages->exist;
}

sub initialize {
	my ( $self ) = @_;
	die "there exist already a repository at ".$self->real_dir if $self->is_initialized;
	$self->mailrc->save;
	$self->packages->save;
	$self->perms->save;
}

sub add_author_distribution {
	my ( $self, $author, $distribution_filename, $path ) = @_;
	my @fileparts = splitdir( $distribution_filename );
	my $filename = pop(@fileparts);
	my $author_path_filename;
	my $target_dir = $self->mkauthordir($author);
	if ($path) {
		my $path_dir = catfile( $self->splitted_dir, $self->authorbase_path_parts, $path );
		$self->mkdir( $path_dir ) unless -d $path_dir;
		$author_path_filename = catfile( $path, $filename );
	} else {
		$author_path_filename = catfile( $self->author_path_parts($author), $filename );
	}
	copy($distribution_filename,catfile( $self->splitted_dir, $self->authorbase_path_parts, $author_path_filename ));
	$self->packages->add_distribution($author_path_filename)->save;
	$self->mailrc->set_alias($author)->save unless defined $self->mailrc->aliases->{$author};
	return catfile( $self->authorbase_path_parts, $self->author_path_parts($author), $filename );
}

sub set_perms {
	my $self = shift;
	$self->perms->set_perms(@_)->save;
}

sub set_alias {
	my ( $self, $author, $alias ) = @_;
	$self->mailrc->set_alias($author,$alias)->save;
}

sub mkauthordir {
	my ( $self, $author ) = @_;
	my $authordir = $self->authordir($author);
	$self->mkdir( $authordir ) unless -d $authordir;
	return $authordir;
}

sub author_path_parts {
	my ( $self, $author ) = @_;
	return substr( $author, 0, 1 ), substr( $author, 0, 2 ), $author;
}

sub authorbase_path_parts { 'authors', 'id' }

sub authordir {
	my ( $self, $author ) = @_;
	return catdir( $self->splitted_dir, $self->authorbase_path_parts, $self->author_path_parts($author) );
}

sub modules {
	my ( $self ) = @_;
	my %modules;
	for (keys %{$self->packages->modules}) {
		$modules{$_} = catfile( $self->splitted_dir, $self->authorbase_path_parts, splitdir( $self->packages->modules->{$_}->[1] ) );
	}
	return \%modules;
}

sub timestamp { shift->packages->timestamp }

#
# Utilities
#

sub mkdir {
	my ( $self, @path ) = @_;
	make_path(catdir(@path),{ error => \my $err });
	if (@$err) {
		for my $diag (@$err) {
			my ($file, $message) = %$diag;
			if ($file eq '') {
				die "general error: $message\n";
			} else {
				die "problem making path $file: $message\n";
			}
		}
	}
}

1;


__END__
=pod

=head1 NAME

CPAN::Repository - API to access a directory which can be served as CPAN repository

=head1 VERSION

version 0.008

=head1 SYNOPSIS

  use CPAN::Repository;

  my $repo = CPAN::Repository->new({
    dir => '/var/www/greypan.company.org/htdocs/',
    url => 'http://greypan.company.org/',
  });
  
  $repo->initialize unless $repo->is_initialized;
  
  $repo->add_author_distribution('AUTHOR','My-Distribution-0.001.tar.gz');
  $repo->add_author_distribution('AUTHOR2','Other-Dist-0.001.tar.gz','Custom/Own/Path');
  $repo->set_alias('AUTHOR','The author <author@company.org>');
  $repo->set_alias('AUTHOR2','The other author <author@company.org>');
  
  my %modules = %{$repo->modules};
  
  my $fullpath_to_authordir = $self->authordir('SOMEONE');
 
  my $packages = $self->packages; # gives back a CPAN::Repository::Packages
  my $mailrc = $self->mailrc; # gives back a CPAN::Repository::Mailrc

=head1 DESCRIPTION

This module is made for representing a directory which can be used as own CPAN for modules, so it can be a GreyPAN, a DarkPAN or even can be
used to manage a mirror of real CPAN your own way. Some code parts are taken from CPAN::Dark of B<CHROMATIC> and L<CPAN::Mini::Inject> of B<MITHALDU>.

=encoding utf8

=head1 SEE ALSO

L<CPAN::Repository::Packages>

L<CPAN::Repository::Mailrc>

=head1 SUPPORT

IRC

  Join #duckduckgo on irc.freenode.net. Highlight Getty for fast reaction :).

Repository

  http://github.com/Getty/p5-cpan-repository
  Pull request and additional contributors are welcome

Issue Tracker

  http://github.com/Getty/p5-cpan-repository/issues

=head1 AUTHOR

Torsten Raudssus <torsten@raudss.us> L<http://raudss.us/>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by DuckDuckGo, Inc. L<http://duckduckgo.com/>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/CPAN/Repository/Mailrc.pm
package CPAN::Repository::Mailrc;
BEGIN {
  $CPAN::Repository::Mailrc::AUTHORITY = 'cpan:GETTY';
}
{
  $CPAN::Repository::Mailrc::VERSION = '0.008';
}
# ABSTRACT: 01mailrc

use Moo;

with qw(
	CPAN::Repository::Role::File
);

sub file_parts { 'authors', '01mailrc.txt' }

has aliases => (
	is => 'ro',
	lazy => 1,
	builder => '_build_aliases',
);

sub _build_aliases {
	my ( $self ) = @_;
	return {} unless $self->exist;
	my @lines = $self->get_file_lines;
	my %aliases;
	for (@lines) {
		if ($_ =~ m/^alias (\w+) "(.*)"$/) {
			$aliases{$1} = $2;
		}
	}
	return \%aliases;
}

sub set_alias {
	my ( $self, $author, $alias ) = @_;
	$self->aliases->{$author} = $alias;
	return $self;
}

sub get_alias {
	my ( $self, $author ) = @_;
	return defined $self->aliases->{$author}
		? $self->aliases->{$author}
		: ();
}

sub generate_content {
	my ( $self ) = @_;
	my $content = "";
	for (sort { $a cmp $b } keys %{$self->aliases}) {
		$content .= 'alias '.$_.' "'.( $self->aliases->{$_} ? $self->aliases->{$_} : $_ ).'"'."\n";
	}
	return $content;
}

1;



__END__
=pod

=head1 NAME

CPAN::Repository::Mailrc - 01mailrc

=head1 VERSION

version 0.008

=head1 SYNOPSIS

  use CPAN::Repository::Mailrc;

  my $mailrc = CPAN::Repository::Mailrc->new({
    repository_root => $fullpath_to_root,
  });

=encoding utf8

=head1 SEE ALSO

L<CPAN::Repository>

=head1 SUPPORT

IRC

  Join #duckduckgo on irc.freenode.net. Highlight Getty for fast reaction :).

Repository

  http://github.com/Getty/p5-cpan-repository
  Pull request and additional contributors are welcome

Issue Tracker

  http://github.com/Getty/p5-cpan-repository/issues

=head1 AUTHOR

Torsten Raudssus <torsten@raudss.us> L<http://raudss.us/>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by DuckDuckGo, Inc. L<http://duckduckgo.com/>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/CPAN/Repository/Packages.pm
package CPAN::Repository::Packages;
BEGIN {
  $CPAN::Repository::Packages::AUTHORITY = 'cpan:GETTY';
}
{
  $CPAN::Repository::Packages::VERSION = '0.008';
}
# ABSTRACT: 02packages

use Moo;

with qw(
	CPAN::Repository::Role::File
);

use Dist::Data;
use File::Spec::Functions ':ALL';
use IO::File;
use DateTime::Format::RFC3339;
use DateTime::Format::Epoch::Unix;

sub file_parts { 'modules', '02packages.details.txt' }
sub file_parts_stamp { 'modules', '02STAMP' }

has modules => (
	is => 'ro',
	lazy => 1,
	builder => '_build_modules',
);

sub _build_modules {
	my ( $self ) = @_;
	return {} unless $self->exist;
	my @lines = $self->get_file_lines;
	my %modules;
	for (@lines) {
		chomp($_);
		next if ($_ =~ /[^:]:[ \t]/);
		if ($_ =~ m/^([^ \t]+)[ \t]+([^ \t]+)[ \t]+([^ \t]+)$/) {
			# $1 = module
			# $2 = version
			# $3 = path (inside repository)
			$modules{$1} = [ $2, $3 ];
		}
	}
	return \%modules;
}

has authorbase_path_parts => (
	is => 'ro',
	required => 1,
);

has url => (
	is => 'ro',
	required => 1,
);

has written_by => (
	is => 'ro',
	required => 1,
);

sub get_module {
	my ( $self, $module ) = @_;
	return defined $self->modules->{$module}
		? $self->modules->{$module}->[1]
		: ();
}

sub get_module_version {
	my ( $self, $module ) = @_;
	return defined $self->modules->{$module}
		? $self->modules->{$module}->[0]
		: ();
}

sub set_module {
	my ( $self, $module, $version, $path ) = @_;
	return $self->modules->{$module} = [ $version, $path ];
}

sub add_distribution {
	my ( $self, $author_distribution_path ) = @_;
	my $filename = catfile( $self->repository_root, @{$self->authorbase_path_parts}, splitdir( $author_distribution_path ) );
	my $dist = Dist::Data->new( $filename );
	for (keys %{$dist->packages}) {
		$self->set_module($_, $dist->packages->{$_}->{version}, $author_distribution_path);
	}
	return $self;
}

sub stamp_filename {
	my ( $self ) = @_;
	catfile( $self->repository_root, $self->file_parts_stamp );
}

after save => sub {
	my ( $self ) = @_;
	my $stamp = IO::File->new($self->stamp_filename, "w") or die "cant write to ".$self->stamp_filename;
	my $now = DateTime->now;
	print $stamp (DateTime::Format::Epoch::Unix->format_datetime($now).' '.DateTime::Format::RFC3339->new->format_datetime($now)."\n");
	$stamp->close;
};

sub timestamp {
	my ( $self ) = @_;
	my $stamp = IO::File->new($self->stamp_filename, "r") or die "cant read ".$self->stamp_filename;
	my ( $line ) = <$stamp>;
	chomp($line);
	if ($line =~ /^(\d+) /) {
		return DateTime::Format::Epoch::Unix->parse_datetime($1);
	} else {
		die "cant find unix timestamp from ".$self->stamp_filename;
	}
}

sub generate_content {
	my ( $self ) = @_;
	my @file_parts = $self->file_parts;
	my $content = "";
	$content .= $self->generate_header_line('File:',(pop @file_parts));
	$content .= $self->generate_header_line('URL:',$self->url.$self->path_inside_root);
	$content .= $self->generate_header_line('Description:','Package names found in directory $CPAN/authors/id/');
	$content .= $self->generate_header_line('Columns:','package name, version, path');
	$content .= $self->generate_header_line('Intended-For:','Automated fetch routines, namespace documentation.');
	$content .= $self->generate_header_line('Written-By:',$self->written_by);
	$content .= $self->generate_header_line('Line-Count:',scalar keys %{$self->modules});
	$content .= $self->generate_header_line('Last-Updated:',DateTime->now->strftime('%a, %e %b %y %T %Z'));
	$content .= "\n";
	for (sort { $a cmp $b } keys %{$self->modules}) {
		$content .= sprintf("%-60s %-20s %s\n",$_,$self->modules->{$_}->[0] ? $self->modules->{$_}->[0] : 'undef',$self->modules->{$_}->[1]);
	}
	return $content;
}

sub generate_header_line {
	my ( $self, $key, $value ) = @_;
	return sprintf("%-13s %s\n",$key,$value);
}

1;


__END__
=pod

=head1 NAME

CPAN::Repository::Packages - 02packages

=head1 VERSION

version 0.008

=head1 SYNOPSIS

  use CPAN::Repository::Packages;

  my $packages = CPAN::Repository::Packages->new({
    repository_root => $fullpath_to_root,
    url => $url,
    written_by => $written_by,
    authorbase_path_parts => ['authors','id'],
  });

=encoding utf8

=head1 SEE ALSO

L<CPAN::Repository>

=head1 SUPPORT

IRC

  Join #duckduckgo on irc.freenode.net. Highlight Getty for fast reaction :).

Repository

  http://github.com/Getty/p5-cpan-repository
  Pull request and additional contributors are welcome

Issue Tracker

  http://github.com/Getty/p5-cpan-repository/issues

=head1 AUTHOR

Torsten Raudssus <torsten@raudss.us> L<http://raudss.us/>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by DuckDuckGo, Inc. L<http://duckduckgo.com/>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/CPAN/Repository/Role/File.pm
package CPAN::Repository::Role::File;
BEGIN {
  $CPAN::Repository::Role::File::AUTHORITY = 'cpan:GETTY';
}
{
  $CPAN::Repository::Role::File::VERSION = '0.008';
}
# ABSTRACT: Role for file functions

use Moo::Role;

use File::Path qw( make_path );
use File::Spec::Functions ':ALL';
use IO::Zlib;
use IO::File;

requires qw( file_parts generate_content );

has repository_root => (
	is => 'ro',
	required => 1,
);

has generate_uncompressed => (
	is => 'ro',
	lazy => 1,
	builder => '_build_generate_uncompressed',
);

sub _build_generate_uncompressed { 1 }

sub path_inside_root {
	my ( $self ) = @_;
	return join("/",$self->file_parts);
}

sub compressed_path_inside_root {
	my ( $self ) = @_;
	return join("/",$self->file_parts).".gz";
}

sub full_filename {
	my ( $self ) = @_;
	return catfile( splitdir($self->repository_root), $self->file_parts );
}

sub full_compressed_filename { shift->full_filename.".gz" }

sub exist {
	my ( $self ) = @_;
	return 0 unless -f $self->full_compressed_filename;
	return 1;
}

sub save {
	my ( $self ) = @_;
	my @pps = $self->file_parts;
	pop(@pps);
	$self->mkdir( splitdir( $self->repository_root ), @pps ) unless -d catdir( $self->repository_root, @pps );
	my $content = $self->generate_content;
	my $gz = IO::Zlib->new($self->full_compressed_filename, "w") or die "cant write to ".$self->full_compressed_filename;
	print $gz $content;
	$gz->close;
	if ($self->generate_uncompressed) {
		my $txt = IO::File->new($self->full_filename, "w") or die "cant write to ".$self->full_filename;
		print $txt $content;
		$txt->close;
	}
	return 1;
}

sub get_file_lines {
	my ( $self ) = @_;
	my $gz = IO::Zlib->new($self->full_compressed_filename, "r") or die "cant read ".$self->full_compressed_filename;
	return <$gz>;
}

sub mkdir {
	my ( $self, @path ) = @_;
	make_path(catdir(@path),{ error => \my $err });
	if (@$err) {
		for my $diag (@$err) {
			my ($file, $message) = %$diag;
			if ($file eq '') {
				die "general error: $message\n";
			} else {
				die "problem making path $file: $message\n";
			}
		}
	}
}

1;


__END__
=pod

=head1 NAME

CPAN::Repository::Role::File - Role for file functions

=head1 VERSION

version 0.008

=encoding utf8

=head1 SUPPORT

IRC

  Join #duckduckgo on irc.freenode.net. Highlight Getty for fast reaction :).

Repository

  http://github.com/Getty/p5-cpan-repository
  Pull request and additional contributors are welcome

Issue Tracker

  http://github.com/Getty/p5-cpan-repository/issues

=head1 AUTHOR

Torsten Raudssus <torsten@raudss.us> L<http://raudss.us/>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by DuckDuckGo, Inc. L<http://duckduckgo.com/>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/CPAN/Repository/Perms.pm
package CPAN::Repository::Perms;
BEGIN {
  $CPAN::Repository::Perms::AUTHORITY = 'cpan:GETTY';
}
{
  $CPAN::Repository::Perms::VERSION = '0.008';
}
# ABSTRACT: 06perms

use Moo;

with qw(
	CPAN::Repository::Role::File
);

use Dist::Data;
use File::Spec::Functions ':ALL';
use IO::File;
use DateTime::Format::RFC3339;
use DateTime::Format::Epoch::Unix;

sub file_parts { 'modules', '06perms.txt' }

our @perm_perms = qw( m f c );

has perms => (
	is => 'rw',
	lazy => 1,
	builder => '_build_perms',
);

sub _build_perms {
	my ( $self ) = @_;
	return {} unless $self->exist;
	my @lines = $self->get_file_lines;
	my %perms;
	for (@lines) {
		chomp($_);
		next if ($_ =~ /[^:]:[ \t]/);
		next if ($_ =~ /^ +/);
		next unless $_;
		my @perm_parts = split(',',$_);
		if (@perm_parts == 3) {
			my $module = shift @perm_parts;
			my $userid = shift @perm_parts;
			my $perm = shift @perm_parts;
			$perms{$module} = {} unless defined $perms{$module};
			warn "already found entry for ".$userid." on ".$module if defined $perms{$module}->{$userid};
			$perms{$module}->{$userid} = $perm;
		}
	}
	return \%perms;
}

has written_by => (
	is => 'ro',
	required => 1,
);

sub perms_linecount {
	my ( $self ) = @_;
	my $i = 0;
	for (keys %{$self->perms}) {
		for my $u (keys %{$self->perms->{$_}}) {
			$i++;
		}
	}
	return $i;
}

sub set_perms {
	my ( $self, $module, $userid, $perm ) = @_;
	die "unknown perm ".$perm unless grep { $_ eq $perm } @perm_perms;
	$self->perms->{$module} = {} unless defined $self->perms->{$module};
	$self->perms->{$module}->{$userid} = $perm;
	return $self;
}

sub get_perms {
	my ( $self, $module ) = @_;
	return defined $self->perms->{$module}
		? $self->perms->{$module}
		: {}
}

sub get_perms_by_userid {
	my ( $self, $userid ) = @_;
	my %perms_by_userid;
	for (keys %{$self->perms}) {
		if (defined $self->perms->{$_}->{$userid}) {
			$perms_by_userid{$_} = $self->perms->{$_}->{$userid};
		}
	}
	return \%perms_by_userid;
}

# File:        06perms.txt
# Description: CSV file of upload permission to the CPAN per namespace
#     best-permission is one of "m" for "modulelist", "f" for
#     "first-come", "c" for "co-maint"
# Columns:     package,userid,best-permission
# Line-Count:  215301
# Written-By:  PAUSE version 1.14
# Date:        Fri, 06 Jul 2012 20:23:21 GMT

sub generate_content {
	my ( $self ) = @_;
	my @file_parts = $self->file_parts;
	my $content = "";
	$content .= $self->generate_header_line('File:',(pop @file_parts));
	$content .= $self->generate_header_line('Description:','Description: CSV file of upload permission to the CPAN per namespace');
	$content .= '    best-permission is one of "m" for "modulelist", "f" for'."\n";
	$content .= '    "first-come", "c" for "co-maint"'."\n";
	$content .= $self->generate_header_line('Columns:','package, userid, best-permission');
	$content .= $self->generate_header_line('Intended-For:','Automated fetch routines, namespace documentation.');
	$content .= $self->generate_header_line('Written-By:',$self->written_by);
	$content .= $self->generate_header_line('Line-Count:',$self->perms_linecount);
	$content .= $self->generate_header_line('Date:',DateTime->now->strftime('%a, %e %b %y %T %Z'));
	$content .= "\n";
	for (sort { $a cmp $b } keys %{$self->perms}) {
		for my $u (sort { $a cmp $b } keys %{$self->perms->{$_}}) {
			$content .= join(",",$_,$u,$self->perms->{$_}->{$u});
		}
	}
	return $content;
}

sub generate_header_line {
	my ( $self, $key, $value ) = @_;
	return sprintf("%-13s %s\n",$key,$value);
}

1;


__END__
=pod

=head1 NAME

CPAN::Repository::Perms - 06perms

=head1 VERSION

version 0.008

=head1 SYNOPSIS

  use CPAN::Repository::Perms;

  my $packages = CPAN::Repository::Perms->new({
    repository_root => $fullpath_to_root,
    written_by => $written_by,
  });

=encoding utf8

=head1 SEE ALSO

L<CPAN::Repository>

=head1 SUPPORT

IRC

  Join #duckduckgo on irc.freenode.net. Highlight Getty for fast reaction :).

Repository

  http://github.com/Getty/p5-cpan-repository
  Pull request and additional contributors are welcome

Issue Tracker

  http://github.com/Getty/p5-cpan-repository/issues

=head1 AUTHOR

Torsten Raudssus <torsten@raudss.us> L<http://raudss.us/>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by DuckDuckGo, Inc. L<http://duckduckgo.com/>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut

